utilities/transactions/write_prepared_txn.cc

```diff
@@ -91,34 +91,29 @@ Status WritePreparedTxn::CommitWithoutPrepareInternal() {
   return CommitBatchInternal(GetWriteBatch()->GetWriteBatch());
 }
 
-SequenceNumber WritePreparedTxn::GetACommitSeqNumber(SequenceNumber prep_seq) {
-  if (db_impl_->immutable_db_options().two_write_queues) {
-    auto s = db_impl_->IncAndFetchSequenceNumber();
-#ifndef NDEBUG
-    MutexLock l(&wpt_db_->seq_for_metadata_mutex_);
-    wpt_db_->seq_for_metadata.push_back(s);
-#endif
-    return s;
-  } else {
-    return prep_seq;
-  }
-}
-
 Status WritePreparedTxn::CommitBatchInternal(WriteBatch* batch) {
   // TODO(myabandeh): handle the duplicate keys in the batch
   // In the absence of Prepare markers, use Noop as a batch separator
   WriteBatchInternal::InsertNoop(batch);
-  const bool disable_memtable = true;
+  const bool DISABLE_MEMTABLE = true;
   const uint64_t no_log_ref = 0;
   uint64_t seq_used = kMaxSequenceNumber;
   auto s = db_impl_->WriteImpl(write_options_, batch, nullptr, nullptr,
-                               no_log_ref, !disable_memtable, &seq_used);
+                               no_log_ref, !DISABLE_MEMTABLE, &seq_used);
   assert(seq_used != kMaxSequenceNumber);
   uint64_t& prepare_seq = seq_used;
-  uint64_t commit_seq = GetACommitSeqNumber(prepare_seq);
-  // TODO(myabandeh): skip AddPrepared
-  wpt_db_->AddPrepared(prepare_seq);
-  wpt_db_->AddCommitted(prepare_seq, commit_seq);
+  // Commit the batch by writing an empty batch to the queue that will release
+  // the commit sequence number to readers.
+  WritePreparedCommitEntryPreReleaseCallback update_commit_map(
```

Comment:
Wny not provide a callback to the first WriteImpl to release the sequence?

Reply:
The release of the sequence MUST be done only from one write queue otherwise we will have gaps in published sequences. When we use two write queues, the 2nd write will go the 2nd queue, which is the designated queue for publishing the sequence number. When using one queue, both writes go to the same queue.

Reply:
Got it. Now that WriteBatchInternal() is no much difference with Prepare+Commit. Why not go through normal transaction to write it?

Reply:
The user might do that. We are penalizing CommitBatch in favor of 2PC transactions so they are no longer a cheaper option under WritePreparedTxn. So the users who switch to WritePreparedTxn might consider to use Prepare/Commit sequence. Note that there might be other reason for not using Prepare/Commit sequence. For example if a mysql user does not use binlog, it does not have to go though the group commit of mysql and can directly call CommitBatch.